src/main/kotlin/org/rust/lang/core/psi/ext/PsiElement.kt

```diff
@@ -332,3 +332,20 @@ fun PsiElement.isKeywordLike(): Boolean {
         else -> false
     }
 }
+
+/**
+ * Consider we do some `resolve` in a quick-fix which is called in preview mode.
+ * Quick-fix is called on copy of the original file, but `resolve` will return original element.
+ * We will have an exception if we try to modify the original element.
+ * Thus, we should call this function on `resolve` result to obtain element in the copy of the original file.
+ */
+fun <T: PsiElement> T.findPreviewCopy(startElement: PsiElement): T? {
+    if (!startElement.isIntentionPreviewElement) return this
+    val copyFile = startElement.containingFile
+    return when (containingFile) {
+        copyFile -> this
+        copyFile.originalFile -> PsiTreeUtil.findSameElementInCopy(this, copyFile)
+        // TODO this is incorrect, intention will fail with "Must not change PSI outside write action"
+        else -> this
```

Comment:
What if we just return `null` here?

Reply:
It depends on when `findPreviewCopyIfNeeded` is used. If it is used at the beginning of the quick-fix, the preview will be empty. If it used somewhere in the middle we may end up in partial preview generated